core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -83,7 +87,9 @@ public void processExpiration() {
             // We limit it so there is only so many async expiration removals done at the same time
             List<CompletableFuture> futures = new ArrayList<>(MAX_ASYNC_EXPIRATIONS);
             long currentTimeMillis = timeService.wallClockTime();
-            dataContainer.running().forEachIncludingExpired((ice, segment) -> {
+            LocalizedCacheTopology topology = distributionManager.getCacheTopology();
+            IntSet segments = IntSets.from(topology.getReadConsistentHash().getPrimarySegmentsForOwner(localAddress));
+            dataContainer.running().forEachIncludingExpired(segments, (ice, segment) -> {
```

Comment:
Could/should we stop processing expiration when the topology changes?
Detecting the change is really cheap, just `distributionManager.getCacheTopology() != topology`

Reply:
We can add that, sure.